test/lib/PlacesTestUtils.js

```diff
@@ -24,22 +26,69 @@ const PlacesTestUtils = Object.freeze({
    *            [optional] visitDate: visit date in microseconds from the epoch
    *            [optional] referrer: nsIURI of the referrer for this visit
    *          }
-   *
-   * @return {Promise}
-   *            resolves When all visits have been added successfully.
-   *            rejects JavaScript exception.
    */
   addVisits: Task.async(function*(placeInfo) {
-    let promise = new Promise((resolve, reject) => {
-      let places = [];
-      if (placeInfo instanceof Ci.nsIURI) {
-        places.push({uri: placeInfo});
-      } else if (Array.isArray(placeInfo)) {
-        places = places.concat(placeInfo);
-      } else {
-        places.push(placeInfo);
+    let places = [];
+
+    if (placeInfo instanceof Ci.nsIURI) {
+      places.push({uri: placeInfo});
+    } else if (Array.isArray(placeInfo)) {
+      places = places.concat(placeInfo);
+    } else {
+      places.push(placeInfo);
+    }
+
+    // Keep track of the number of rows for final validation
+    function* getPlacesRows() {
```

Comment:
Ok, but I'm still not sure we want to use `yield*` here. It's going to be mega confusing (not just because of the naming, but because `yield*` injects itself mid-step creating a weird "Inception"... yes, like in the movie Inception!). 

Using Task.spawn() is cleaner IMHO, because it let's you know straight away that:

 1. it returns a promise, on which the other generator can yield (see my code again). 
 1. it immediately runs the generator.
 1. it doesn't require the generator itself to be called, or for it to even have a name - it is just passed as an argument and Task.spawn() does the right thing.  

Re: the naming... We should try to named things like "function* generateFoos()" or "function* asyncFooRunner()"... to make it more clear what a generator does.